pkg/connector/tickets.go

```diff
@@ -0,0 +1,449 @@
+	var ret []*v2.TicketSchema
+
+	offset := 0
+	// get offset from page token if its not empty
+	if p != nil && p.Token != "" {
+		var err error
+		offset, err = strconv.Atoi(p.Token)
+		if err != nil {
+			return nil, "", nil, err
+		}
+	}
+
+	projects, _, err := j.client.Project.Find(ctx, jira.WithStartAt(offset), jira.WithMaxResults(p.Size), jira.WithExpand("issueTypes"))
+	if err != nil {
+		return nil, "", nil, wrapError(err, "failed to get projects")
+	}
+
+	for _, project := range projects {
+		schema, err := j.schemaForProject(ctx, project)
+		if err != nil {
+			return nil, "", nil, err
+		}
+		ret = append(ret, schema)
+	}
+
+	nextPageToken := ""
+	if len(ret) > 1 {
+		nextPageToken = fmt.Sprintf("%d", offset+len(ret))
+	}
+
+	return ret, nextPageToken, nil, nil
+}
+
+func (j *Jira) getTicketStatuses(ctx context.Context, projectId string) ([]*v2.TicketStatus, error) {
+	if j.ticketStatuses != nil {
+		return j.ticketStatuses, nil
+	}
+
+	statuses, _, err := j.client.Status.GetAllStatuses(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// filter statuses by project id
+	var filteredStatuses []jira.Status
+	for _, status := range statuses {
+		if status.Scope != nil && status.Scope.Project != nil && status.Scope.Project.Id == projectId {
+			filteredStatuses = append(filteredStatuses, status)
+		}
+	}
+
+	ret := make([]*v2.TicketStatus, 0, len(filteredStatuses))
+	for _, status := range filteredStatuses {
+		ret = append(ret, &v2.TicketStatus{
+			Id:          status.ID,
+			DisplayName: status.Name,
+		})
+	}
+
+	j.ticketStatuses = ret
+
+	return ret, nil
+}
+
+func (j *Jira) schemaForProject(ctx context.Context, project jira.Project) (*v2.TicketSchema, error) {
+	var ticketTypes []*v2.TicketType
+	customFields := make(map[string]*v2.TicketCustomField)
+
+	var components []*v2.TicketCustomFieldObjectValue
+
+	for _, issueType := range project.IssueTypes {
+		// TODO: Maybe we care about subtasks?
+		if !issueType.Subtask {
+			ticketTypes = append(ticketTypes, &v2.TicketType{
+				Id:          issueType.ID,
+				DisplayName: issueType.Name,
+			})
+		}
+	}
+	for _, component := range project.Components {
+		components = append(components, &v2.TicketCustomFieldObjectValue{
+			Id:          component.ID,
+			DisplayName: component.Name,
+		})
+	}
+
+	// Add a required field for the project
+	customFields["project"] = sdkTicket.PickObjectValueFieldSchema(
+		"project",
+		"Project",
+		true,
+		[]*v2.TicketCustomFieldObjectValue{
+			{
+				Id:          project.ID,
+				DisplayName: project.Name,
+			},
+		},
+	)
+
+	if len(components) > 0 {
+		customFields["components"] = sdkTicket.PickMultipleObjectValuesFieldSchema(
+			"components",
+			"Components",
+			false,
+			components,
+		)
+	}
+
+	ret := &v2.TicketSchema{
+		Id:           project.Key,
+		DisplayName:  project.Name,
+		Types:        ticketTypes,
+		CustomFields: customFields,
+	}
+
+	statuses, err := j.getTicketStatuses(ctx, project.ID)
```

Comment:
remove this and replace by filtering the passed in list to global or project id matches

do all these and I think baton-jira is done?